### ISSUE TITLE
Errata/Package install requests via API result in Fault 2107 ModulesNotAllowedException

### ISSUE THREAD

**Issue by santeri3700** [uyuni-project/uyuni#6429]
### Problem description

Calling `system.schedulePackageInstall` or `system.scheduleApplyErrata` via the XML-RPC or REST API will always result in Fault 2107 unless `allowModules` parameter is set to true.

I've been able to reproduce the issue with CentOS 7, AlmaLinux 8, Ubuntu 18.04 LTS and Debian 10 clients. Also to my knowledge, CentOS 7 and Debian-based distributions do not have modular repositories in the same sense as EL8 has so this error message doesn't make sense.

The Web UI does not show the module warning message unless a system is actually subscribed to channels with modular repositories (e.g. AlmaLinux 8 without CLM). The API error is the same regardless what is visible in the Web UI (modules present or not).

This behavior has started with Uyuni 2022.11 and is present in 2022.12 too. Tested with two separate 2022.12 servers and one 2022.11 server. Uyuni 2022.10 was/is unaffected as far as I can tell.

Example XML-RPC API response:
```
Fault 2107: 'redstone.xmlrpc.XmlRpcFault: At least one of the channels this system is subscribed to contains modules. If you have activated modules on this system, please refrain from using Uyuni for package operations. Instead, use <a href="/rhn/manager/contentmanagement/projects">Content Lifecycle Management</a> to <strong>create regular channels that include module packages</strong>.'
```

### Steps to reproduce

Tested with CentOS 7, AlmaLinux 8, Ubuntu 18.04 LTS and Debian 10 clients with channels synchronized via `spacewalk-common-channels`.
1. Set up an Uyuni 2022.11 or 2022.12 server
2. Synchronize various channels/repositories & create and build a CLM project for AlmaLinux 8 using the default AppStream filters.
3. Bootstrap client(s) with the mentioned distributions and matching software channels.
4. Inspect the available package and/or errata IDs for a system. These can be either available updates or available packages that are not installed on the system.
5. Call `system.schedulePackageInstall` or `system.scheduleApplyErrata` via the XML-RPC or REST API which will always result in Fault 2107 unless `allowModules` parameter is set to true.

Example REST API calls with cURL without `allowModules` parameter:
```
$ curl -H "Content-Type: application/json" --cookie "pxt-session-cookie=.....;" -X POST "http://uyuni-server-fqdn/rhn/manager/api/system/scheduleApplyErrata" -d "{'sid': '1000010000', 'errataIds': [1234], 'earliestOccurrence': '2032-01-01T12:00:00Z'}"

{"success":false,"message":"At least one of the channels this system is subscribed to contains modules. If you have activated modules on this system, please refrain from using Uyuni for package operations. Instead, use \u003ca href\u003d\"/rhn/manager/contentmanagement/projects\"\u003eContent Lifecycle Management\u003c/a\u003e to \u003cstrong\u003ecreate regular channels that include module packages\u003c/strong\u003e."}
```
```
$ curl -H "Content-Type: application/json" --cookie "pxt-session-cookie=.....;" -X POST "http://uyuni-server-fqdn/rhn/manager/api/system/schedulePackageInstall" -d "{'sid': '1000010000', 'packageIds': [4321], 'earliestOccurrence': '2032-01-01T12:00:00Z'}"

{"success":false,"message":"At least one of the channels this system is subscribed to contains modules. If you have activated modules on this system, please refrain from using Uyuni for package operations. Instead, use \u003ca href\u003d\"/rhn/manager/contentmanagement/projects\"\u003eContent Lifecycle Management\u003c/a\u003e to \u003cstrong\u003ecreate regular channels that include module packages\u003c/strong\u003e."}
```

### Uyuni version

Uyuni 2022.11 server:
```
Information for package Uyuni-Server-release:
---------------------------------------------
Repository     : Uyuni Server Stable
Name           : Uyuni-Server-release
Version        : 2022.11-220400.195.1.uyuni2
Arch           : x86_64
Vendor         : obs://build.opensuse.org/systemsmanagement:Uyuni
Support Level  : Level 3
Installed Size : 1.4 KiB
Installed      : Yes
Status         : out-of-date (version 2022.11-220400.193.4.uyuni2 installed)
Source package : Uyuni-Server-release-2022.11-220400.195.1.uyuni2.src
```

Uyuni 2022.12 servers (actually 2022.12 even though the package info says 2022.11):
```
Information for package Uyuni-Server-release:
---------------------------------------------
Repository     : uyuni-server-stable
Name           : Uyuni-Server-release
Version        : 2022.11-220400.195.1.uyuni2
Arch           : x86_64
Vendor         : obs://build.opensuse.org/systemsmanagement:Uyuni
Support Level  : Level 3
Installed Size : 1.4 KiB
Installed      : Yes
Status         : up-to-date
Source package : Uyuni-Server-release-2022.11-220400.195.1.uyuni2.src
```

### Useful logs

The debug logs don't show much since the Java code does not seem to contain a lot of relevant log calls (even with full debug logging). Only the rhn_web_ui.log seems to show something related to the API calls.

/var/log/rhn/rhn_web_ui.log:
```
2023-01-05 10:50:59,411 [ajp-nio-0:0:0:0:0:0:0:1-8009-exec-8] DEBUG com.redhat.rhn.frontend.servlets.EnvironmentFilter - set REQUESTED_URI: /rhn/manager/api/system/schedulePackageInstall
...
<Lots of authentication related logs lines>
...
2023-01-05 10:50:59,418 [ajp-nio-0:0:0:0:0:0:0:1-8009-exec-8] DEBUG com.redhat.rhn.common.db.datasource.CachedStatement - execute() - Executing: SELECT 1
  FROM rhnPackage P
 WHERE p.id = ?
   AND (   P.org_id = ?
        OR EXISTS (SELECT 1
                     FROM rhnOrgChannelTreeView CT inner join rhnChannelPackage CP on CT.id = CP.channel_id
                    WHERE CT.org_id = ?
                      AND CP.package_id = P.id)
        OR EXISTS (SELECT 1
                     FROM rhnSharedChannelTreeView CT inner join rhnChannelPackage CP on CT.id = CP.channel_id
                    WHERE CT.org_id = ?
                      AND CP.package_id = P.id)
       )
2023-01-05 10:50:59,418 [ajp-nio-0:0:0:0:0:0:0:1-8009-exec-8] DEBUG com.redhat.rhn.common.db.datasource.CachedStatement - execute() - With: {org_id=1, pid=4321}
2023-01-05 10:50:59,422 [ajp-nio-0:0:0:0:0:0:0:1-8009-exec-8] DEBUG com.redhat.rhn.common.db.datasource.CachedStatement - execute() - Return type: true
2023-01-05 10:50:59,427 [ajp-nio-0:0:0:0:0:0:0:1-8009-exec-8] DEBUG com.redhat.rhn.common.db.datasource.CachedStatement - execute() - Executing: SELECT 1
  FROM rhnUserServerPerms USP
 WHERE USP.user_id = ?
   AND USP.server_id = ?
2023-01-05 10:50:59,427 [ajp-nio-0:0:0:0:0:0:0:1-8009-exec-8] DEBUG com.redhat.rhn.common.db.datasource.CachedStatement - execute() - With: {uid=1, sid=1000010000}
2023-01-05 10:50:59,427 [ajp-nio-0:0:0:0:0:0:0:1-8009-exec-8] DEBUG com.redhat.rhn.common.db.datasource.CachedStatement - execute() - Return type: true
2023-01-05 10:50:59,429 [ajp-nio-0:0:0:0:0:0:0:1-8009-exec-8] DEBUG com.redhat.rhn.common.localization.messages - Resolving message "packagelist.jsp.modulespresent" for locale en_US
2023-01-05 10:50:59,429 [ajp-nio-0:0:0:0:0:0:0:1-8009-exec-8] DEBUG com.redhat.rhn.common.localization.XmlMessages - Got bundle from cache, returning : com.redhat.rhn.frontend.strings.jsp.StringResource.en_US
```

### Additional information

Some relevant parts of the Java code:
- https://github.com/uyuni-project/uyuni/blob/c6a0d80a/java/code/src/com/redhat/rhn/frontend/xmlrpc/system/SystemHandler.java#L3663-L3672
- https://github.com/uyuni-project/uyuni/blob/c6a0d80a/java/code/src/com/redhat/rhn/frontend/xmlrpc/system/SystemHandler.java#L3971-L3985
- https://github.com/uyuni-project/uyuni/blob/c6a0d80a/java/code/src/com/redhat/rhn/domain/channel/Channel.java#L260-L269
- https://github.com/uyuni-project/uyuni/blob/c6a0d80a/java/code/src/com/redhat/rhn/domain/channel/Channel.java#L280-L282

I'm more than happy to do more testing and gather logs if needed.

**Comment by mbussolotto** [comment]
Hi @santeri3700 , thanks for reporting
IIUC, the workaround is set `allowModules` to `true`, righ ?

**Comment by santeri3700** [comment]
> Hi @santeri3700 , thanks for reporting IIUC, the workaround is set `allowModules` to `true`, righ ?

Yes, indeed. It works with the XML-RPC and REST APIs.

**Comment by admd** [comment]
Should be resolved with https://github.com/uyuni-project/uyuni/pull/6440. Fix will be part of next Uyuni release.

**Comment by admd** [comment]
> Thanks for sharing - I can confirm that this also applies to SUSE Manager 4.3.3 and SLES 12 SP5 clients even though there are no modules. Manually setting `True` might or might not help.

Thank you, Christian. Yes, we are aware of it and fix was back ported to 4.3 as well.

**Comment by santeri3700** [comment]
I've finally tested the fix with Uyuni 2023.01 (XML-RPC & REST) and can confirm it solved the problem. Many thanks to the contributors!
